File: examples/objective-c-http/test_native.m
#import "Transport.h"
#include <arpa/inet.h>
#include <unistd.h>

static BOOL connectPort(int port) {
    int socketFD = socket(AF_INET, SOCK_STREAM, 0);
    if (socketFD < 0) abort();
    struct sockaddr_in address = {.sin_family = AF_INET, .sin_port = htons(port)};
    address.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    int result = connect(socketFD, (struct sockaddr *)&address, sizeof(address));
    close(socketFD);
    return result == 0;
}

int main(void) {
    @autoreleasepool {
        for (NSString *security in @[@"tls", @"e2ee"]) {
            StogasTransportOwner *transport = [[StogasTransportOwner alloc]
                initWithConfiguration:@{@"environment":@"staging", @"security":security}];
            if (!transport) return 1;
            int port = [[[NSURL URLWithString:transport.baseURL] port] intValue];
            if (!connectPort(port)) return 1;
            @try {
                if ([security isEqual:@"e2ee"]) @throw [NSException exceptionWithName:@"Interrupted" reason:nil userInfo:nil];
            } @catch (NSException *error) { (void)error; }
            @finally { [transport close]; [transport close]; }
            if (connectPort(port)) return 1;
#if !__has_feature(objc_arc)
            [transport release];
#endif
        }
        StogasTransportOwner *invalid = [[StogasTransportOwner alloc]
            initWithConfiguration:@{@"security":@"unknown"}];
        if (invalid) return 1;
        puts("Native ownership and exception cleanup passed");
        return 0;
    }
}

File: examples/zig-http/test_native.zig
const std = @import("std");
const native = @import("transport.zig");
const socket = @cImport({
    @cInclude("arpa/inet.h");
    @cInclude("unistd.h");
});

fn connect(port: u16) !void {
    const fd = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0);
    if (fd < 0) return error.SocketFailed;
    defer _ = socket.close(fd);
    var address: socket.sockaddr_in = std.mem.zeroes(socket.sockaddr_in);
    address.sin_family = socket.AF_INET;
    address.sin_port = socket.htons(port);
    address.sin_addr.s_addr = socket.htonl(0x7f000001);
    if (socket.connect(fd, @ptrCast(&address), @sizeOf(socket.sockaddr_in)) != 0)
        return error.ConnectionFailed;
}

fn useTransport(port: *u16, interrupted: bool) !void {
    var transport = try native.Transport.init(std.testing.allocator, if (interrupted)
        "{\"environment\":\"staging\",\"security\":\"e2ee\"}"
    else
        "{\"environment\":\"staging\",\"security\":\"tls\"}");
    defer transport.deinit();
    port.* = (try std.Uri.parse(transport.base_url)).port orelse return error.MissingPort;
    try connect(port.*);
    if (interrupted) return error.Interrupted;
}

test "native transport releases its listener on normal return and error" {
    var port: u16 = 0;
    try useTransport(&port, false);
    try std.testing.expectError(error.ConnectionFailed, connect(port));
    try std.testing.expectError(error.Interrupted, useTransport(&port, true));
    try std.testing.expectError(error.ConnectionFailed, connect(port));
    try std.testing.expectError(error.TransportUnavailable, native.Transport.init(std.testing.allocator, "{\"security\":\"unknown\"}"));
}
